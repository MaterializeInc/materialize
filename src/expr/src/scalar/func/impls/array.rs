// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::{ColumnType, Datum, RowArena, ScalarType};

use crate::scalar::func::{stringify_datum, LazyUnaryFunc};
use crate::{EvalError, MirScalarExpr};

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastArrayToListOneDim;

impl LazyUnaryFunc for CastArrayToListOneDim {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }

        let arr = a.unwrap_array();
        let ndims = arr.dims().ndims();
        if ndims > 1 {
            return Err(EvalError::Unsupported {
                feature: format!(
                    "casting multi-dimensional array to list; got array with {} dimensions",
                    ndims
                ),
                issue_no: None,
            });
        }

        Ok(Datum::List(arr.elements()))
    }

    /// The output ColumnType of this function
    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        ScalarType::List {
            element_type: Box::new(input_type.scalar_type.unwrap_array_element_type().clone()),
            custom_id: None,
        }
        .nullable(true)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        true
    }

    fn inverse(&self) -> Option<crate::UnaryFunc> {
        None
    }
}

impl fmt::Display for CastArrayToListOneDim {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("arraytolist")
    }
}

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastArrayToString {
    pub ty: ScalarType,
}

impl LazyUnaryFunc for CastArrayToString {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }
        let mut buf = String::new();
        stringify_datum(&mut buf, a, &self.ty)?;
        Ok(Datum::String(temp_storage.push_string(buf)))
    }

    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        ScalarType::String.nullable(input_type.nullable)
    }

    fn propagates_nulls(&self) -> bool {
        true
    }

    fn introduces_nulls(&self) -> bool {
        false
    }

    fn preserves_uniqueness(&self) -> bool {
        true
    }

    fn inverse(&self) -> Option<crate::UnaryFunc> {
        // TODO? If we moved typeconv into `expr` we could determine the right
        // inverse of this.
        None
    }
}

impl fmt::Display for CastArrayToString {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("arraytostr")
    }
}
