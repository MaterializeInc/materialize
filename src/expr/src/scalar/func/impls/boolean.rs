// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use mz_repr::strconv;

sqlfunc!(
    #[sqlname = "NOT"]
    #[preserves_uniqueness = true]
    fn not(a: bool) -> bool {
        !a
    }
);

sqlfunc!(
    #[sqlname = "boolean_to_text"]
    #[preserves_uniqueness = true]
    fn cast_bool_to_string<'a>(a: bool) -> &'a str {
        match a {
            true => "true",
            false => "false",
        }
    }
);

sqlfunc!(
    #[sqlname = "boolean_to_nonstandard_text"]
    #[preserves_uniqueness = true]
    fn cast_bool_to_string_nonstandard<'a>(a: bool) -> &'a str {
        // N.B. this function differs from `cast_bool_to_string` because
        // the SQL specification requires `true` and `false` to be spelled out in
        // explicit casts, while PostgreSQL prefers its more concise `t` and `f`
        // representation in some contexts, for historical reasons.
        strconv::format_bool_static(a)
    }
);

sqlfunc!(
    #[sqlname = "boolean_to_integer"]
    #[preserves_uniqueness = true]
    fn cast_bool_to_int32(a: bool) -> i32 {
        match a {
            true => 1,
            false => 0,
        }
    }
);
