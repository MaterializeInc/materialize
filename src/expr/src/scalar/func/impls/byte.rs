// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use mz_repr::strconv;

use crate::EvalError;

sqlfunc!(
    #[sqlname = "bytea_to_text"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastStringToBytes)]
    fn cast_bytes_to_string(a: &'a [u8]) -> String {
        let mut buf = String::new();
        strconv::format_bytes(&mut buf, a);
        buf
    }
);

sqlfunc!(
    #[sqlname = "bit_length"]
    fn bit_length_bytes<'a>(a: &'a [u8]) -> Result<i32, EvalError> {
        let val = a.len() * 8;
        i32::try_from(val).or(Err(EvalError::Int32OutOfRange(val.to_string())))
    }
);

sqlfunc!(
    #[sqlname = "octet_length"]
    fn byte_length_bytes<'a>(a: &'a [u8]) -> Result<i32, EvalError> {
        let val = a.len();
        i32::try_from(val).or(Err(EvalError::Int32OutOfRange(val.to_string())))
    }
);
