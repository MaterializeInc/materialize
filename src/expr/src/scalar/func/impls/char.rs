// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::adt::char::{format_str_pad, Char, CharLength};
use mz_repr::{ColumnType, ScalarType};

use crate::scalar::func::EagerUnaryFunc;

/// All Char data is stored in Datum::String with its blank padding removed
/// (i.e. trimmed), so this function provides a means of restoring any
/// removed padding.
#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct PadChar {
    pub length: Option<CharLength>,
}

impl<'a> EagerUnaryFunc<'a> for PadChar {
    type Input = &'a str;
    type Output = Char<String>;

    fn call(&self, a: &'a str) -> Char<String> {
        Char(format_str_pad(a, self.length))
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Char {
            length: self.length,
        }
        .nullable(input.nullable)
    }
}

impl fmt::Display for PadChar {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("padchar")
    }
}

// This function simply allows the expression of changing a's type from varchar to string
sqlfunc!(
    #[sqlname = "char_to_text"]
    #[preserves_uniqueness = true]
    fn cast_char_to_string<'a>(a: Char<&'a str>) -> &'a str {
        a.0
    }
);
