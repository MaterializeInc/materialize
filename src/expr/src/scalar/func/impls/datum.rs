// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use mz_repr::{Datum, DatumList};

use crate::EvalError;

sqlfunc!(
    #[sqlname = "isnull"]
    fn is_null<'a>(a: Datum<'a>) -> bool {
        a.is_null()
    }
);

sqlfunc!(
    #[sqlname = "istrue"]
    fn is_true<'a>(a: Datum<'a>) -> bool {
        a == Datum::True
    }
);

sqlfunc!(
    #[sqlname = "isfalse"]
    fn is_false<'a>(a: Datum<'a>) -> bool {
        a == Datum::False
    }
);

sqlfunc!(
    fn pg_column_size<'a>(a: Datum<'a>) -> Result<Option<i32>, EvalError> {
        match a {
            Datum::Null => Ok(None),
            datum => {
                let sz = mz_repr::datum_size(&datum);
                i32::try_from(sz)
                    .map(Some)
                    .or(Err(EvalError::Int32OutOfRange))
            }
        }
    }
);

sqlfunc!(
    fn mz_row_size<'a>(a: DatumList<'a>) -> Result<i32, EvalError> {
        let sz = mz_repr::row_size(a.iter());
        i32::try_from(sz).or(Err(EvalError::Int32OutOfRange))
    }
);
