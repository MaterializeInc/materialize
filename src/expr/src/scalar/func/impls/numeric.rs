// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use dec::{OrderedDecimal, Rounding};
use repr::adt::numeric::{self, Numeric};
use repr::strconv;

use crate::EvalError;

sqlfunc!(
    #[sqlname = "-"]
    #[preserves_uniqueness = true]
    fn neg_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().neg(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "abs"]
    fn abs_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().abs(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "ceilnumeric"]
    fn ceil_numeric(mut a: Numeric) -> Numeric {
        // ceil will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Ceiling);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "expnumeric"]
    fn exp_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.exp(&mut a);
        let cx_status = cx.status();
        if cx_status.overflow() {
            Err(EvalError::FloatOverflow)
        } else if cx_status.subnormal() {
            Err(EvalError::FloatUnderflow)
        } else {
            numeric::munge_numeric(&mut a).unwrap();
            Ok(a)
        }
    }
);

sqlfunc!(
    #[sqlname = "floornumeric"]
    fn floor_numeric(mut a: Numeric) -> Numeric {
        // floor will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Floor);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

fn log_guard_numeric(val: &Numeric, function_name: &str) -> Result<(), EvalError> {
    if val.is_negative() {
        return Err(EvalError::NegativeOutOfDomain(function_name.to_owned()));
    }
    if val.is_zero() {
        return Err(EvalError::ZeroOutOfDomain(function_name.to_owned()));
    }
    Ok(())
}

// From the `decNumber` library's documentation:
// > Inexact results will almost always be correctly rounded, but may be up to 1
// > ulp (unit in last place) in error in rare cases.
//
// See decNumberLog10 documentation at http://speleotrove.com/decimal/dnnumb.html
fn log_numeric<F>(mut a: Numeric, logic: F, name: &'static str) -> Result<Numeric, EvalError>
where
    F: Fn(&mut dec::Context<Numeric>, &mut Numeric),
{
    log_guard_numeric(&a, name)?;
    let mut cx = numeric::cx_datum();
    logic(&mut cx, &mut a);
    numeric::munge_numeric(&mut a).unwrap();
    Ok(a)
}

sqlfunc!(
    #[sqlname = "lnnumeric"]
    fn ln_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::ln, "ln")
    }
);

sqlfunc!(
    #[sqlname = "log10numeric"]
    fn log10_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::log10, "log10")
    }
);

sqlfunc!(
    #[sqlname = "roundnumeric"]
    fn round_numeric(mut a: Numeric) -> Numeric {
        // round will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        numeric::cx_datum().round(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "sqrtnumeric"]
    fn sqrt_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        if a.is_negative() {
            return Err(EvalError::NegSqrt);
        }
        let mut cx = numeric::cx_datum();
        cx.sqrt(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        Ok(a)
    }
);

sqlfunc!(
    #[sqlname = "numerictoi16"]
    fn cast_numeric_to_int16(mut a: Numeric) -> Result<i16, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        let i = cx.try_into_i32(a).or(Err(EvalError::Int16OutOfRange))?;
        i16::try_from(i).or(Err(EvalError::Int16OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi32"]
    fn cast_numeric_to_int32(mut a: Numeric) -> Result<i32, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i32(a).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi64"]
    fn cast_numeric_to_int64(mut a: Numeric) -> Result<i64, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i64(a).or(Err(EvalError::Int64OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictof32"]
    fn cast_numeric_to_float32(a: Numeric) -> Result<f32, EvalError> {
        let i = a.to_string().parse::<f32>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float32OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictof64"]
    fn cast_numeric_to_float64(a: Numeric) -> Result<f64, EvalError> {
        let i = a.to_string().parse::<f64>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float64OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictostr"]
    fn cast_numeric_to_string(a: Numeric) -> String {
        let mut buf = String::new();
        strconv::format_numeric(&mut buf, &OrderedDecimal(a));
        buf
    }
);
