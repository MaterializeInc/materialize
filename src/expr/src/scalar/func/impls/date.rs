// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use chrono::{DateTime, Datelike, NaiveDate, NaiveDateTime, Utc};
use mz_repr::adt::date::Date;
use mz_repr::adt::timestamp::{CheckedTimestamp, DateLike};
use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::adt::datetime::DateTimeUnits;
use mz_repr::adt::numeric::Numeric;
use mz_repr::{strconv, ColumnType, ScalarType};

use crate::scalar::func::EagerUnaryFunc;
use crate::EvalError;

sqlfunc!(
    #[sqlname = "date_to_text"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastStringToDate)]
    fn cast_date_to_string(a: Date) -> String {
        let mut buf = String::new();
        strconv::format_date(&mut buf, a);
        buf
    }
);

sqlfunc!(
    #[sqlname = "date_to_timestamp"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastTimestampToDate)]
    fn cast_date_to_timestamp(a: Date) -> Result<CheckedTimestamp<NaiveDateTime>, EvalError> {
        Ok(CheckedTimestamp::from_timestamplike(
            NaiveDate::from(a).and_hms(0, 0, 0),
        )?)
    }
);

sqlfunc!(
    #[sqlname = "date_to_timestamp_with_timezone"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastTimestampTzToDate)]
    fn cast_date_to_timestamp_tz(a: Date) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
        Ok(CheckedTimestamp::from_timestamplike(
            DateTime::<Utc>::from_utc(NaiveDate::from(a).and_hms(0, 0, 0), Utc),
        )?)
    }
);

pub fn extract_date_inner(units: DateTimeUnits, date: NaiveDate) -> Result<Numeric, EvalError> {
    match units {
        DateTimeUnits::Epoch => Ok(Numeric::from(date.extract_epoch())),
        DateTimeUnits::Millennium => Ok(Numeric::from(date.millennium())),
        DateTimeUnits::Century => Ok(Numeric::from(date.century())),
        DateTimeUnits::Decade => Ok(Numeric::from(date.decade())),
        DateTimeUnits::Year => Ok(Numeric::from(date.year())),
        DateTimeUnits::Quarter => Ok(Numeric::from(date.quarter())),
        DateTimeUnits::Week => Ok(Numeric::from(date.iso_week_number())),
        DateTimeUnits::Month => Ok(Numeric::from(date.month())),
        DateTimeUnits::Day => Ok(Numeric::from(date.day())),
        DateTimeUnits::DayOfWeek => Ok(Numeric::from(date.day_of_week())),
        DateTimeUnits::DayOfYear => Ok(Numeric::from(date.ordinal())),
        DateTimeUnits::IsoDayOfWeek => Ok(Numeric::from(date.iso_day_of_week())),
        DateTimeUnits::Hour
        | DateTimeUnits::Minute
        | DateTimeUnits::Second
        | DateTimeUnits::Milliseconds
        | DateTimeUnits::Microseconds => Err(EvalError::UnsupportedUnits(
            format!("{}", units),
            "date".to_string(),
        )),
        DateTimeUnits::Timezone
        | DateTimeUnits::TimezoneHour
        | DateTimeUnits::TimezoneMinute
        | DateTimeUnits::IsoDayOfYear => Err(EvalError::Unsupported {
            feature: format!("'{}' timestamp units", units),
            issue_no: None,
        }),
    }
}

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct ExtractDate(pub DateTimeUnits);

impl<'a> EagerUnaryFunc<'a> for ExtractDate {
    type Input = Date;
    type Output = Result<Numeric, EvalError>;

    fn call(&self, a: Date) -> Result<Numeric, EvalError> {
        extract_date_inner(self.0, a.into())
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Numeric { max_scale: None }.nullable(input.nullable)
    }
}

impl fmt::Display for ExtractDate {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "extract_{}_d", self.0)
    }
}
