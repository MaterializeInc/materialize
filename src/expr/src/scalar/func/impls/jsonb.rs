// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::adt::jsonb::{Jsonb, JsonbRef};
use mz_repr::adt::numeric::{self, Numeric, NumericMaxScale};
use mz_repr::{strconv, ColumnType, Datum, Row, RowPacker, ScalarType};

use crate::scalar::func::impls::numeric::*;
use crate::scalar::func::EagerUnaryFunc;
use crate::EvalError;

sqlfunc!(
    #[sqlname = "jsonb_to_text"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastStringToJsonb)]
    fn cast_jsonb_to_string<'a>(a: JsonbRef<'a>) -> String {
        let mut buf = String::new();
        strconv::format_jsonb(&mut buf, a);
        buf
    }
);

sqlfunc!(
    #[sqlname = "jsonb_to_smallint"]
    fn cast_jsonb_to_int16<'a>(a: JsonbRef<'a>) -> Result<i16, EvalError> {
        match a.into_datum() {
            Datum::Numeric(a) => cast_numeric_to_int16(a.into_inner()),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "smallint".into(),
            }),
        }
    }
);

sqlfunc!(
    #[sqlname = "jsonb_to_integer"]
    fn cast_jsonb_to_int32<'a>(a: JsonbRef<'a>) -> Result<i32, EvalError> {
        match a.into_datum() {
            Datum::Numeric(a) => cast_numeric_to_int32(a.into_inner()),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "integer".into(),
            }),
        }
    }
);

sqlfunc!(
    #[sqlname = "jsonb_to_bigint"]
    fn cast_jsonb_to_int64<'a>(a: JsonbRef<'a>) -> Result<i64, EvalError> {
        match a.into_datum() {
            Datum::Numeric(a) => cast_numeric_to_int64(a.into_inner()),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "bigint".into(),
            }),
        }
    }
);

sqlfunc!(
    #[sqlname = "jsonb_to_real"]
    fn cast_jsonb_to_float32<'a>(a: JsonbRef<'a>) -> Result<f32, EvalError> {
        match a.into_datum() {
            Datum::Numeric(a) => cast_numeric_to_float32(a.into_inner()),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "real".into(),
            }),
        }
    }
);

sqlfunc!(
    #[sqlname = "jsonb_to_double"]
    fn cast_jsonb_to_float64<'a>(a: JsonbRef<'a>) -> Result<f64, EvalError> {
        match a.into_datum() {
            Datum::Numeric(a) => cast_numeric_to_float64(a.into_inner()),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "double precision".into(),
            }),
        }
    }
);

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastJsonbToNumeric(pub Option<NumericMaxScale>);

impl<'a> EagerUnaryFunc<'a> for CastJsonbToNumeric {
    type Input = JsonbRef<'a>;
    type Output = Result<Numeric, EvalError>;

    fn call(&self, a: JsonbRef<'a>) -> Result<Numeric, EvalError> {
        match a.into_datum() {
            Datum::Numeric(mut num) => match self.0 {
                None => Ok(num.into_inner()),
                Some(scale) => {
                    if numeric::rescale(&mut num.0, scale.into_u8()).is_err() {
                        return Err(EvalError::NumericFieldOverflow);
                    };
                    Ok(num.into_inner())
                }
            },
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "numeric".into(),
            }),
        }
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Numeric { max_scale: self.0 }.nullable(input.nullable)
    }
}

impl fmt::Display for CastJsonbToNumeric {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("jsonb_to_numeric")
    }
}

sqlfunc!(
    #[sqlname = "jsonb_to_boolean"]
    fn cast_jsonb_to_bool<'a>(a: JsonbRef<'a>) -> Result<bool, EvalError> {
        match a.into_datum() {
            Datum::True => Ok(true),
            Datum::False => Ok(false),
            datum => Err(EvalError::InvalidJsonbCast {
                from: jsonb_typeof(JsonbRef::from_datum(datum)).into(),
                to: "boolean".into(),
            }),
        }
    }
);

sqlfunc!(
    #[sqlname = "jsonb_or_null_to_jsonb"]
    fn cast_jsonb_or_null_to_jsonb<'a>(a: Option<JsonbRef<'a>>) -> JsonbRef<'a> {
        match a.map(|v| v.into_datum()) {
            None => JsonbRef::from_datum(Datum::JsonNull),
            Some(Datum::Numeric(n)) => {
                let n = n.into_inner();
                let datum = if n.is_finite() {
                    Datum::from(n)
                } else if n.is_nan() {
                    Datum::String("NaN")
                } else if n.is_negative() {
                    Datum::String("-Infinity")
                } else {
                    Datum::String("Infinity")
                };
                JsonbRef::from_datum(datum)
            }
            Some(datum) => JsonbRef::from_datum(datum),
        }
    }
);

sqlfunc!(
    fn jsonb_array_length<'a>(a: JsonbRef<'a>) -> Result<Option<i32>, EvalError> {
        match a.into_datum() {
            Datum::List(list) => match i32::try_from(list.iter().count()) {
                Ok(len) => Ok(Some(len)),
                Err(_) => Err(EvalError::Int32OutOfRange),
            },
            _ => Ok(None),
        }
    }
);

sqlfunc!(
    fn jsonb_typeof<'a>(a: JsonbRef<'a>) -> &'a str {
        match a.into_datum() {
            Datum::Map(_) => "object",
            Datum::List(_) => "array",
            Datum::String(_) => "string",
            Datum::Numeric(_) => "number",
            Datum::True | Datum::False => "boolean",
            Datum::JsonNull => "null",
            d => panic!("Not jsonb: {:?}", d),
        }
    }
);

sqlfunc!(
    fn jsonb_strip_nulls<'a>(a: JsonbRef<'a>) -> Jsonb {
        fn strip_nulls(a: Datum, row: &mut RowPacker) {
            match a {
                Datum::Map(dict) => row.push_dict_with(|row| {
                    for (k, v) in dict.iter() {
                        match v {
                            Datum::JsonNull => (),
                            _ => {
                                row.push(Datum::String(k));
                                strip_nulls(v, row);
                            }
                        }
                    }
                }),
                Datum::List(list) => row.push_list_with(|row| {
                    for elem in list.iter() {
                        strip_nulls(elem, row);
                    }
                }),
                _ => row.push(a),
            }
        }
        let mut row = Row::default();
        strip_nulls(a.into_datum(), &mut row.packer());
        Jsonb::from_row(row)
    }
);

sqlfunc!(
    fn jsonb_pretty<'a>(a: JsonbRef<'a>) -> String {
        let mut buf = String::new();
        strconv::format_jsonb_pretty(&mut buf, a);
        buf
    }
);
