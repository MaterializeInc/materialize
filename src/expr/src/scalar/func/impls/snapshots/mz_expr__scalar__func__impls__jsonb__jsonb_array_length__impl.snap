---
source: src/expr/src/scalar/func/impls/jsonb.rs
expression: info
---
Info {
    output_type_nullable: ColumnType {
        scalar_type: Int32,
        nullable: true,
    },
    output_type_nonnullable: ColumnType {
        scalar_type: Int32,
        nullable: true,
    },
    preserves_uniqueness: false,
    inverse: None,
    is_monotone: false,
    propagates_nulls: true,
    introduces_nulls: true,
    could_error: true,
}
